sensitivity-analysis/src/main/java/com/farao_community/farao/sensitivity_analysis/ra_sensi_handler/InjectionRangeActionSensiHandler.java

```diff
@@ -0,0 +1,114 @@
+        List<SensitivityVariable> sensitivityVariables = new ArrayList<>();
+
+        /*
+         separate injection distribution keys of the RangeAction into two GlskMap:
+         - one with the positive injection variations
+         - one with the negative injection variations
+
+         Rescale the GlskMaps so that their sum is equal to 1
+         */
+
+        Map<String, Float> positiveGlskMap = getPositiveGlskMap();
+        Map<String, Float> negativeGlskMap = getNegativeGlskMap();
+
+        if (!positiveGlskMap.isEmpty()) {
+            rescaleGlskMap(positiveGlskMap);
+            sensitivityVariables.add(new LinearGlsk(getPositiveGlskMapId(), getPositiveGlskMapId(), positiveGlskMap));
+        }
+
+        if (!negativeGlskMap.isEmpty()) {
+            rescaleGlskMap(negativeGlskMap);
+            sensitivityVariables.add(new LinearGlsk(getNegativeGlskMapId(), getNegativeGlskMapId(), negativeGlskMap));
+        }
+
+        return sensitivityVariables;
+    }
+
+    @Override
+    public double getSensitivityOnFlow(FlowCnec cnec, SystematicSensitivityResult sensitivityResult) {
+        return sensitivityResult.getSensitivityOnFlow(getPositiveGlskMapId(), cnec) * getKeySum(getPositiveGlskMap())
+                - sensitivityResult.getSensitivityOnFlow(getNegativeGlskMapId(), cnec) * getKeySum(getNegativeGlskMap());
+    }
+
+    @Override
+    public void checkConsistency(Network network) {
+        injectionRangeAction.getInjectionDistributionKeys().forEach((k, v) -> {
+            Identifiable<?> identifiable = network.getIdentifiable(k.getId());
+            if (!(identifiable instanceof Generator || identifiable instanceof Load)) {
+                throw new SensitivityAnalysisException(String.format("Unable to create sensitivity variable for InjectionRangeAction %s, on element %s", injectionRangeAction.getId(), k.getId()));
+            }
+        });
+    }
+
+    private Map<String, Float> getPositiveGlskMap() {
+        return injectionRangeAction.getInjectionDistributionKeys().entrySet()
+                .stream().filter(e -> e.getValue() > 0)
+                .collect(Collectors.toMap(e -> e.getKey().getId(), e -> e.getValue().floatValue()));
+    }
+
+    private Map<String, Float> getNegativeGlskMap() {
+        return injectionRangeAction.getInjectionDistributionKeys().entrySet()
+                .stream().filter(e -> e.getValue() < 0)
+                .collect(Collectors.toMap(e -> e.getKey().getId(), e -> -e.getValue().floatValue()));
+    }
+
+    private void rescaleGlskMap(Map<String, Float> glskMap) {
+        float keySum = (float) getKeySum(glskMap);
+        if (keySum != 0) {
+            glskMap.entrySet().forEach(e -> e.setValue(e.getValue() / keySum));
+        }
+    }
+
+    private double getKeySum(Map<String, Float> glskMap) {
+        return glskMap.values().stream()
+                .mapToDouble(v -> v)
+                .sum();
```

Comment:
maybe summing then converting to double would be less consuming

Reply:
the sum method is only available for a DoubleStream, and the DoubleStream is created with mapToDouble().